Allow user to specify columns in status CLI

knelli2 opened the pull request:
## Proposed changes

Now a user can choose from a preselected list of available columns which ones they want displayed. This can also be set in the config file so it's automatically taken care of.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Allow user to specify columns in status CLI

knelli2 commented on tools/Status/Status.py:
```diff
@@ -206,32 +215,83 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+    elif field == "SegmentId":
+        if pd.isnull(value):
+            return "-"
+        else:
+            return f"Segment_{int(value):04}"
```
I know this needs to be done better, but I couldn't figure out how to get the number of digits from the Segment class.

nilsvu replied:
I'd say move the `pd.isnull` down to the catch-all `else` clause. The segment number can just render as an integer, no need to prefix `Segment_`.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -506,9 +555,39 @@ def render_status(
     """Gives an overview of simulations running on this machine."""
 
+    if available_columns:
+        rich.print(f"Available columns: {list(_available_fields().keys())}")
```
```suggestion
        rich.print(rich.columns.Columns(_available_fields().keys()))
```

No need to print "Available columns: ". The user knows what's going on because they explicitly specified `--available-columns`.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -206,32 +215,83 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
         return str(value)
 
 
+def _available_fields():
```
Static constants like this are implemented in Python by convention as all-caps module-level variables:

```py
AVAILABLE_FIELDS = { ... }
```

Also:
- It would be helpful to name this constant and the two new CLI options consistently as "columns" (instead of "fields") so it's clear they belong together.
- Please add a comment about what the key and the value mean (one is the Slurm field and the other is the column name).
- I suggest you use this dictionary to do a `job_data.rename(columns=AVAILABLE_COLUMNS)`. Then you don't have to generate the columns names manually. For this you have to switch the key and value.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -506,9 +555,39 @@ def render_status(
+    "--columns",
+    "-c",
+    "input_columns",
+    cls=PythonLiteralOption,
```
Just take the columns as a comma-separated list:

```py
@click.option(
     "--columns",
     "-c",
     callback=(
        lambda ctx, param, value: (
            value.split(",") if value else []
        )
    ),
)
```

nilsvu reviewed (changes requested):
Please squash directly

knelli2 commented:
Added a fixup with the changes. I also split the changes for adding additional columns to a separate commit so now the original commit/fixup keeps the same available (and default) columns.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
-        else:
-            return humanize.naturaldate(value) + " " + value.strftime("%X")
+        return humanize.naturaldate(value) + " " + value.strftime("%X")
+    elif field == "Segment":
+        return str(int(value)).zfill(4)
```
Just delete so the number prints as an int (no need for the zfill)

knelli2 replied:
But I want 4 digits to match the actual segment. I think it's easier to view if it's consistent with what's after the underscore of the segment. I'd like to leave this as is unless you have very strong objections.

nilsvu replied:
Ok use `Segment.NUM_DIGITS` then

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+# is the human-readable name. These are sometimes the same
+AVAILABLE_COLUMNS = {
+    "State": "State",
+    "Partition": "Partition",
```
Print as "Queue"? That's the name of the option in `spectre schedule` (maybe it should be "partition" for consistency with slurm though).

knelli2 replied:
I thought about using Queue. I don't have a strong preference other than consistency. If we want to keep the schedule option as Queue then I'll make this Queue.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+    for key in _default_columns().values():
+        keys += f"{key},"
+    keys = keys[:-1]
+    return keys
```
This is just `",".join(DEFAULT_COLUMNS)` (yes, Python is easier than C++)

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
     show_deleted,
     show_all_segments,
     state_styles,
+    input_columns,
```
Just name this `standard_columns` then you can drop some of the changes below.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+    return keys
+
+
+def _verify_columns(input_columns: Sequence[str]):
```
I think just delete this function and move the assert to the callback below. Let the user reorder the columns if they want :)

nilsvu commented on tools/Status/Status.py:
```diff
@@ -506,9 +553,35 @@ def render_status(
+@click.option(
+    "--columns",
+    "-c",
+    "input_columns",
```
`standard_columns` (`input` is not a very useful name)

knelli2 replied:
Actually I think `input` here is more descriptive than `standard`. And also more accurate. These are columns that the user passes to the command as "inputs". There's no reason these passed in columns are "standard" by any metric (unless they use the defaults)

nilsvu replied:
you can't and shouldn't prefix every option or function argument with "input" just because it's a "user input".

I'm not attached to "standard". I chose that name to distinguish from the "extra" columns that each executable adds ("standard" = same for each exec). Probably just "columns" is fine too (but a bit misleading because more columns are added per exec).

knelli2 replied:
How about "user_columns"

nilsvu replied:
The intention is to distinguish from the exec-specific columns. I have no better idea than "standard" or "generic". Just plain "columns" is also fine.

nilsvu replied:
`columns_to_use` seems no better than just `columns` 😅

nilsvu replied:
I feel like I'm getting too pedantic here though so if you prefer that name for some reason then that's fine with me.

knelli2 replied:
I prefer the name that gets this merged the fastest lol. I think `columns_to_use` is good enough for this.

nilsvu replied:
Change it to `columns` then. It's public API people set in a config file and without a reason for making it different from the `--columns` option name it's unnecessarily confusing.

knelli2 replied:
done

nilsvu commented on tools/Status/Status.py:
```diff
@@ -506,9 +553,35 @@ def render_status(
+        " specified in the config file with the name 'input_columns'. Note that"
+        " if the '--allusers'  option is not specified, then the \"User\""
+        " column will be omitted. Specify the columns as a comma separated list"
+        " wrapped in single quotes; e.g. 'Col1, Col2, Col3'"
```
You don't need quotes if you remove the spaces (user should know that they need quotes if they put spaces on the command line)

knelli2 replied:
I'll add some docs explaining. I'll keep both possibilities though

nilsvu replied:
Ok. Use real col names I think.

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,92 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+}
+
+
+def _default_columns():
```
Just make this a hardcoded list `DEFAULT_COLUMNS`

nilsvu commented on tools/Status/Status.py:
```diff
@@ -190,7 +190,9 @@ def _state_order(state):
 
 def _format(field: str, value: Any, state_styles: dict) -> str:
-    if field == "State":
+    if pd.isnull(value) or value == "None assigned":
```
Which field can print as "None assigned"? Could you add a comment?

knelli2 replied:
The `NodeList` slurm filed returns "None assigned" if the job is pending. I'll add a comment

nilsvu commented:
You can squash everything together

knelli2 committed:
Add more available columns to status CLI

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,64 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+    "Comment": "Comment",
+}
+
+DEFAULT_COLUMNS = {
```
just a list of col names, no need to make this a dict

nilsvu commented on tools/Status/Status.py:
```diff
@@ -202,36 +204,64 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
+
+    for input_key in result:
+        if input_key not in AVAILABLE_COLUMNS.values():
+            raise ValueError(
```
`ValueError` -> `click.BadParameter`

nilsvu commented on tools/Status/Status.py:
```diff
@@ -213,13 +216,16 @@ def _format(field: str, value: Any, state_styles: dict) -> str:
     "Elapsed": "Elapsed",
-    "NCPUS": "Cores",
-    "NNodes": "Nodes",
+    "NCPUS": "NumCores",
+    "NNodes": "NumNodes",
```
Note that there's limited horizontal space in the terminal and the "Num" prefix makes the col name a bit longer. maybe keep the shorter "Cores" and "Nodes" and make the one below "NodeList" like in Slurm? Cores and nodes are short integers so the col width is set by the col name, whereas "NodeList" is longer so a longer col name doesn't limit the col width.
